### PR TITLE
[IMP] base, web, website, *: differentiate essential & optional cookies

### PR DESCRIPTION
**Pull request by bso-odoo**
*: im_livechat, survey, utm, website_crm_iap_reveal, website_forum,
   website_livechat, website_sale, website_sale_comparison

Before this commit all cookies were considered essential.

This commit makes some of them optional. It also makes it possible for
the website visitor to only accept the essential cookies.

Explanation of the rational and history of this cookies bar improvement here: https://github.com/odoo/odoo/pull/95673#discussion_r960657519

task-2800976

Co-authored-by: Benoit Socias <bso@odoo.com>

### CODE REVIEW

**Comment by rdeodoo** [comment]
We would like to ship this one before the freeze, I have a few PRs of my own to finish before reviewing this one, so if you could rebase a last time it would be great :) (I see there is conflict)

**Comment by bso-odoo** [comment]
> We would like to ship this one before the freeze, I have a few PRs of my own to finish before reviewing this one, so if you could rebase a last time it would be great :) (I see there is conflict)

@rdeodoo Ok, but note that we are still waiting for the general fix about the popups - in order to remove the WIP commit from this PR.

**Comment by rdeodoo** [comment]
We will merge it without it I guess if not ready before the freeze.

File: addons/website/models/website.py
Comment (mart-e) [inline addons/website/models/website.py:88]: That's not asking for consent, that's annoying your users for no valid reasons. Ask or dont show anything.

Comment (pierre-pde) [reply addons/website/models/website.py:88]: Hi @mart-e , 'Only I agree' is the only option available at the moment. 
We prefer to keep it because it was deemed important at some point.

Comment (rdeodoo) [reply addons/website/models/website.py:88]: Note that this PR wasn't reviewed at all yet, it is based on a ~3 years old branch we were not allowed to merge back then, so I am just speaking about the 2 cookies bar mode.

> That's not asking for consent, that's annoying your users for no valid reasons.

We all agree about that.

It's not a matter of how you me or anyone else in Odoo see it but how our users do.

Thought I would summarize the rationale and history behind all this, as I think many Odoo people will be unhappy with this change and would prefer the cookies bar to simply disappear, not being "improved".
We are on the same page and would prefer an internet without cookies bar, we just realized that it was important to provide a solution for those people.

Our user base is requesting it again and again.
That was already the case a few years ago when we managed to convince FP to introduce a very light cookies bar despite it being useless and "people should know better than the GDPR". That's also why we have that weird looking confirm dialog when you enable the settings, it was the compromise back then to be allowed to merge it.

As a reminder, we were having hundreds of tickets back then, always telling people to introduce custom script and use https://cookie-bar.eu/ because "Odoo is smarter than GDPR, you don't need it, please read article XXX of the GDPR, they are wrong, you should involve a lawyer in case of problem, you don't need to comply to it".

I agree that we should try and educate people (as we do with the confirm modal), but the most important part is to satisfy the demand, when they are recurrent and there is no work around, especially when our concurrent does it.

Fact is, we shouldn't even care if this is correct or not regarding GDPR and cookies law policy. People don't know anything about it, they just know (or think) that they should have a cookie bar.
After all, your bank website, your town website, your government website, the police website, everyone has it.
It just became a standard (a shitty one we all agree but that's not the point), and people just want to (easily and quickly) be able to set one on their website, they don't care much more than that.

This one is just going a step further as there is still plenty of people not happy with the current cookies bar (while most are happy with it) and want it to be more granular.

The idea behind keeping both mode (the current one) was to keep a default cookie bar really light and minimalist, with a single "I agree" button and not impacting the business code (no need to check if cookie is set).
But indeed, we should probably just remove that new settings and somehow keep the default cookie bar layout as it is right now, with just an extra "Only essential" link :+1:** 
The less settings the better, and apart from having a lighter/simpler cookies bar, it doesn't really provide anything else, so making the new one also minimalist should be enough.

New mode (current PR):
image
Old mode (without this PR):
image

What we could do (no mode settings):
image

At the end, it is exactly the same issue as for the Google Fonts and the people being sued in Germany, we didn't want to do anything at first, but we will introduce a workaround before end of the month for people to be able to download fonts locally and not use google server. Not point leaving people unhappy and thinking we are wrong. Especially when people are now sending scam attempt to website using google server for fonts. That would not be helping the product.

Comment (mart-e) [reply addons/website/models/website.py:88]: > Note that this PR wasn't reviewed at all yet, it is based on a ~3 years old branch we were not allowed to merge back then

Sorry if commented too fast :) If this still need some work, maybe put it back in draft? I quickly tested and actually got analytic cookies stored _before_ I got to make a choice.

Thank you for the detailed explanation of the history (that I was not aware of). I don't like them but I get why we add the feature to have a cookie bar if people are asking for it. But let's offer a valid one then.
I think the "improved" cookie bar is the only valid one. The `Only "I accept"` should **not** be an option as it does not bring any advantage, the only two choices should be "No banner" or "Choice banner".
Also, the "Only essential" option should not be displayed lightly, in a way to favour the "Agree all", that's a shitty move that we see way too often...

File: addons/website_forum/static/src/js/website_forum.js
Comment (mart-e) [inline addons/website_forum/static/src/js/website_forum.js:505]: Missing `ttl` value so the `optional` is not kept. This makes me think, wouldn't it make sense to check the type of `ttl` in `set_cookie` (detect if not falsy or integer) to avoid similar issues at other places ?

File: addons/website/models/website.py
Comment (bso-odoo) [reply addons/website/models/website.py:88]: Agreed with PO: now only enabling GA only if consent is not an option or if consent was given.

File: addons/website/static/src/js/http_cookie.js
Comment (rdeodoo) [reply addons/website/static/src/js/http_cookie.js:18]: Don't we also say "Pas de nouvelles, bonnes nouvelles ?" :innocent: :trollface:

Comment (bso-odoo) [reply addons/website/static/src/js/http_cookie.js:18]: So, if we do not display a cookie bar at all, it should always be false ? :stuck_out_tongue: 
Just kidding, we can easily get the info of whether or not the question is asked - I would use that information instead.
But for that I'll need a green light from the requirements side which explictly ask for the opposite - and was re-confirmed this morning.

File: addons/website/models/ir_http.py
Comment (mart-e) [inline addons/website/models/ir_http.py:385]: The logic of this method is a bit odd, you call the super but you never check the value and assume `result` is `True`.
Either you take the lead and don't call super at all, or you change a bit to call `super` if you are not in a website-supported case ("required" and "optional"). Let's assume I create a module that adds the type `performance`, this method would force it to `required`.

File: addons/utm/models/ir_http.py
Comment (xmo-odoo) [inline addons/utm/models/ir_http.py:19]: These are not just optional cookies, they are tracking cookies. And most of the optional cookies are not flagged e.g. totp device cookies and frontend lang cookies are not technically required for odoo to *work* (although the totp device cookie has its own opt-in so probably counts as required once the user asks for it). Possibly whatever crm_reveal does as well, I've no idea.

File: addons/website/models/ir_http.py
Comment (xmo-odoo) [inline addons/website/models/ir_http.py:395]: Why is `required` part of the conversation when it's already handled above?

File: addons/website/models/ir_http.py
Comment (xmo-odoo) [inline addons/website/models/ir_http.py:387]: Why are we using error prone strings when the number of possibilities is strictly enumerated?

File: addons/website/models/website.py
Comment (xmo-odoo) [reply addons/website/models/website.py:88]: > Hi @mart-e , 'Only I agree' is the only option available at the moment.

Why not add an option for the site owner to disable optional cookies? Only set required cookies and show no bar.

File: addons/website/static/src/js/http_cookie.js
Comment (xmo-odoo) [inline addons/website/static/src/js/http_cookie.js:28]: Same as the default value in the `contents` map, this is the wrong default.

File: addons/website/static/src/js/http_cookie.js
Comment (bso-odoo) [reply addons/website/static/src/js/http_cookie.js:18]: @mart-e @xmo-odoo FYI Good news, we just managed to move the spec line - the behavior will become the expected one => save optional cookie only if they were accepted - or if no question is asked (i.e. no cookiebar mode)

Comment (xmo-odoo) [reply addons/website/static/src/js/http_cookie.js:18]: > So, if we do not display a cookie bar at all, it should always be false ? stuck_out_tongue

Well yes, that would in fact be the ethical (and in GDPR jurisdictions the legal) behaviour.

File: addons/website/models/ir_http.py
Comment (bso-odoo) [reply addons/website/models/ir_http.py]: Changed to: `website` now only handles `'optional'` - other modules can handle their own value(s).

**Comment by keshrath** [comment]
@bso-odoo  Is this intended for 16.0? Is this not now mandatory for every website in the EU?

File: addons/web/static/src/legacy/js/core/cookie_utils.js
Comment (rdeodoo) [inline addons/web/static/src/legacy/js/core/cookie_utils.js:45]: What is this line about ? It this in case someone wrongly call `set_cookie()` like `set_cookie('cook_name', 'cook_value', 'optional')` ?
If yes, seems overkill, if not could you explain?

File: addons/web/static/src/legacy/js/core/cookie_utils.js
Comment (rdeodoo) [inline addons/web/static/src/legacy/js/core/cookie_utils.js:31]: Nitpicking: The name is a bit confusing, it sounds like it is actually checking the cookie value somehow.
You can ignore this remark and mark as done if you think it is fine as is, otherwise a better name would be great :+1: 
Something with `can` or `allow` maybe

Edit: or simply as the python name which is way clearer: isAllowedCookies

File: addons/website/static/src/js/http_cookie.js
Comment (rdeodoo) [inline addons/website/static/src/js/http_cookie.js:13]: If the goal of this line is actually checking if the cookies bar setting is disabled on the website, should be commented, not straightforward

Comment (rdeodoo) [inline addons/website/static/src/js/http_cookie.js:13]: Also make this return first in the `ìf()` ?
Won't have to parse the cookies, if the cookies bar is disabled, no need to do all that -> Return true
```js
        if (!document.getElementById('cookies-consent-essential')) {
            // Cookies bar is disabled on this website
            return true;
        }
        const consents = JSON.parse(cookieUtils.get_cookie('accepted_cookie_types') || '{}');
        if ('optional' in consents) {
            return consents['optional'];
        }
```

File: odoo/http.py
Comment (rdeodoo) [inline odoo/http.py:1096]: nitpicking: single quote, same for both next lines

File: odoo/http.py
Comment (rdeodoo) [inline odoo/http.py:1063]: Shouldn't we go with the `def set_cookie(self, *args, **kwargs):` notation and pop the cookies type and then `super().set_cookie(*args, **kwargs)` ?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for taking too long before reviewing it. First review reading the code, didn't yet really test.

Could you retarget 16.0?

Quick stuff about the commit message:

> when considered

=> were considered

>  It also adds the possibility for activating a second type of cookie bar that makes it possible for the website visitor to only accept the essential cookies.

Not true anymore, this need to be adapted as we decided to only keep either a cookies bar or no cookies bar, and if cookies bar enabled there will always be the essential / optional behavior.

> Co-authored-by: kzh <kzh@odoo.com>

You can remove that part TBH

> Co-authored-by: bso <bso@odoo.com>

If I am not mistaken, this should be `Benoit Socias <bso@odoo.com>`, not at all important though. How I usually do it is I just find a PR/commit from that person and add `.patch` to the link and copy paste from there, see [here](https://patch-diff.githubusercontent.com/raw/odoo/odoo/pull/101246.patch).

**Comment by bso-odoo** [comment]
> @bso-odoo Is this intended for 16.0? Is this not now mandatory for every website in the EU?

@keshrath This is indeed intended for 16.0. About your second question, I do not think I am qualified to answer it...

File: addons/website/views/website_templates.xml
Comment (bso-odoo) [reply addons/website/views/website_templates.xml:162]: Yes, but also it takes into account the fact that the consent link might have been removed from the template - thus turning it back to an "inform only" cookies bar. In such a case, no choice is left to the user, it falls back to the old behavior.

File: addons/website/models/website.py
Comment (bso-odoo) [reply addons/website/models/website.py:1271]: This makes it possible to click on a link inside the cookie policy that reopens the cookie bar so that you can change your mind.

File: addons/website/static/src/js/editor/snippets.options.js
Comment (bso-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:2597]: Not anymore indeed, this had to be done when we had the option to chose between with or without consent.

File: addons/utm/models/ir_http.py
Comment (rdeodoo) [reply addons/utm/models/ir_http.py:19]: Note that most of the cookies have been checked by FP and PDE which decided, after long debates, at least for the website related ones, if they were required or not.

While we still don't really agree about some like the website lang, that's as per the spec.
There have been long discussion about this already, but if you feel like this should absolutely be changed, I guess we can yet again discuss that with them.
But AFAIC, we always have to fight to make cookies bar related stuff like this PR move on, so any step further is a win tbh, whether the website lang has to be marked as required or not.

About the crm_reveal one, he is flagged as optional as it should IMO.

About the totp one, that's the only one we are unsure if it should be marked as optional. In doubt, we left it as required. We don't want to mess with this one so close to the release.

Let us know if you think something is to be changed.

**Comment by rdeodoo** [comment]
@robodoo rebase-ff

**Comment by mart-e** [comment]
Could we apply the same style to both buttons?
Screenshot 2022-09-29 at 15-37-32 Home My Website

**Comment by rdeodoo** [comment]
It's totally possible in edit mode, see here: https://watch.screencastify.com/v/1EaknemwXp0cVDGZzmK8

**Comment by rdeodoo** [comment]
More seriously, yes we could I guess. Some people might even want the opposite (link in primary and btn in link).
I think it will look uglier (and weird to have 2 primary btn) but we can :shrug: 
image

It will probably be hard to please everyone tho. But since this can be edited later in edit mode..

**Comment by mart-e** [comment]
Of both smaller, I am not a specialist to make things pretty but my point is to avoid the [dark pattern](https://en.wikipedia.org/wiki/Dark_pattern) we see almost everywehre where we push people to accept everything by default.
As a counter example, [europa.eu](https://europa.eu) is thankfully not doing that.

**Comment by rdeodoo** [comment]
Yeah, I figured you were a bit annoyed by the fact the primary action was to accept all cookies. It's a legit concern.

image

image

image

The last solution looks fine to you? I think this should be the one with most chance of surviving long term (SPEC/PO change).

**Comment by mart-e** [comment]
I prefer the first one (2 round white ones, like on belgium.be) but then again, I am not a specialist to make things pretty

File: odoo/http.py
Comment (rdeodoo) [inline odoo/http.py:1065]: Isn't args[0] always set?

**Comment by rdeodoo** [comment]
@mart-e We will adapt as per your preference, we don't really mind which style is chosen for those button, as long as we have the feature move forward :+1: 

@xmo-odoo I believe all your remarks were processed and/or replied.

As this is one of the last (small) new feature we still have to merge for 16.0, I'd like to ship it tomorrow after a final review and testing.
Should anyone want to have a look, now is the time.

File: odoo/http.py
Comment (bso-odoo) [reply odoo/http.py]: That's also what I thought but then a test failed on runbot because `auth_totp` controller calls it with a named `key`, therefore `key` ends up in the `kwargs` instead. ... now fixing this, to fetch the value correctly. Thank you for raising this.

File: addons/website/static/src/js/http_cookie.js
Comment (rdeodoo) [reply addons/website/static/src/js/http_cookie.js]: Could you apply the same in the python def? (no need to check and parse cookies if cookies bar disabled)
```py
    @classmethod
    def _is_allowed_cookie(cls, cookie_type, cookie_name='undefined'):
        result = super()._is_allowed_cookie(cookie_type, cookie_name)
        if result and cookie_type == 'optional':
            accepted_cookie_types = json_scriptsafe.loads(request.httprequest.cookies.get('accepted_cookie_types', '{}'))
            if 'optional' in accepted_cookie_types:
                return accepted_cookie_types['optional']
            return not request.env['website'].get_current_website().cookies_bar

        # Pass-through if already forbidden for another reason or a type that
        # is not restricted by the website module.
        return result
```

```py
    @classmethod
    def _is_allowed_cookie(cls, cookie_type, cookie_name='undefined'):
        result = super()._is_allowed_cookie(cookie_type, cookie_name)
        if result and cookie_type == 'optional':
            if not request.env['website'].get_current_website().cookies_bar:
                # Cookies bar is disabled on this website
                return True
            accepted_cookie_types = json_scriptsafe.loads(request.httprequest.cookies.get('accepted_cookie_types', '{}'))
            if 'optional' in accepted_cookie_types:
                return accepted_cookie_types['optional']

        # Pass-through if already forbidden for another reason or a type that
        # is not restricted by the website module.
        return result
```

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Looks like you forgot to adapt the actual client codebase ?
https://github.com/odoo-dev/odoo/blame/6b4d86b6ebc3776792fef452eb3b1679f674a433/addons/web/static/src/core/browser/cookie_service.js#L38
The PR is adapting the python side, the legacy code, but not the new wowl code?

File: addons/website/static/src/snippets/s_popup/000.js
Comment (rdeodoo) [inline addons/website/static/src/snippets/s_popup/000.js:164]: How can someone close that modal without going through the button? It should be the only way to close it

File: addons/website/static/src/snippets/s_popup/000.js
Comment (rdeodoo) [inline addons/website/static/src/snippets/s_popup/000.js:145]: We are trying to reduce the used cookies but here we are actually creating a second cookie for the cookie bar. We could regroup both.
Indeed the cookie bar is using the popup mechanism which is setting the popup id as cookie name to flag it as read & closed, in this case `website_cookies_bar` which is the ID of the cookie snippet

Can't we simply move the `_onHideModal()` call first in this method (so that it will actually set the cookie `website_cookies_bar`) and then update that cookie val to `{'required': true, 'optional': acceptOptional}` and just read this cookie in the code? (ofc with a comment next to the cookie template to mention that it's id is used as the name of the cookie so no one will change that id blindly)

Your opinion?

There could even be another way, have a new property (smth like that) set to `true` in the `PopupWidget`, and just make your `_onConsentAllClick` and `_onConsentEssentialClick` change that property to `{'required': true, 'optional': xx}`
Then in the hide_modal of `PopupWidget` simply set the cookie to this property.
That will solve the double cookie issue + remove the duplication of the set_cookie and day computation here

File: addons/website/static/src/snippets/s_popup/000.js
Comment (rdeodoo) [inline addons/website/static/src/snippets/s_popup/000.js:124]: Why not the same as the cookies bar 999 (see its data attribute `data-consents-duration="999"`), 

So no need of this property just inline `this.$el.find('.modal').data('consentsDuration');`

Because right now, if one is accepting the cookies bar, he has 2 cookies, the first about showing or not the cookies bar, set for 999 days, and this one set for 365 days, it means that after a year, his choice about cookies preference will not be saved anymore, but he still won't see the cookies bar.

Another reason to use the same cookie imo (see below remark)

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_popup/000.js
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_popup/000.js:164]: You can close the modal by pressing "escape" - this currently is interpreted by the cookies bar as "I'll respond later" (and you get the popup again on the next page load).

Comment (rdeodoo) [reply addons/website/static/src/snippets/s_popup/000.js:164]: It is not supposed to be the case (as opposed to a regular popup), how did you do that? I just tried in 15 and 16 and your branch and I can't close it with escape regardless of the layout and backdrop option.

File: addons/website_livechat/static/src/legacy/website_livechat_chatbot_test_script.js
Comment (bso-odoo) [reply addons/website_livechat/static/src/legacy/website_livechat_chatbot_test_script.js]: Introduced a `deleteCookie(name)` method to make it more clear we do not want further parameters to be taken into account in those cases.

File: addons/website/static/src/snippets/s_popup/000.js
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_popup/000.js]: Indeed, it seems to not be the case anymore. Strange, that behavior was even agreed on with the PO.
I remove that function (will be in the next push).

File: addons/website/static/src/js/http_cookie.js
Comment (bso-odoo) [reply addons/website/static/src/js/http_cookie.js]: Applied change and added a `return False` at the end of the first `if` block.

File: addons/website/static/src/xml/website.cookies_bar.xml
Comment (rdeodoo) [inline addons/website/static/src/xml/website.cookies_bar.xml:31]: The padding on row can't be edited through the editor, that's why it was on the col, could you please fix that in another pr, so it doesn't block this one?

File: odoo/http.py
Comment (rdeodoo) [inline odoo/http.py:1062]: Why are both implementations of `set_cookie` in this file not written the same way?
On top of not being the same (but doing the same), one is fallbacking on None while the other is fallbacking on 'required' (as it should, as per the js def)

Also, why don't we just do nothing if not allowed, instead of removing it (max_age/expires = 0) ?
I mean, why not, but is it really needed? It's weird to have it behaves differently than the js.

**Review by rdeodoo — APPROVED** [review]
@robodoo r+ Thank you for the PR and changes.

Small leftover remark on the padding declaration in XML regarding the website builder integration and one about the response.set_cookie.

This shouldn't block the merge and can be done afterward, as this one is really late (planed for the freeze, but came last as other PR were prioritized due to model change).

**Comment by rdeodoo** [comment]
@robodoo r+
Rebased, there was a conflict due to https://github.com/odoo/odoo/commit/7d23491b44b7372209057ea2abc565a95710b316. I also squashed the one line commit about google analytics in the main one.

File: addons/website/static/src/xml/website.cookies_bar.xml
Comment (bso-odoo) [reply addons/website/static/src/xml/website.cookies_bar.xml:31]: @rdeodoo Moved padding back to column in https://github.com/odoo/odoo/pull/101859.

File: odoo/http.py
Comment (bso-odoo) [reply odoo/http.py:1062]: Unifying both python and the js implementation in https://github.com/odoo/odoo/pull/101859/commits/99b5b753b61383d356675687cb6cd67e88557d1f or https://github.com/odoo/odoo/pull/101859.

**Pull request closed without merging**
